contracts/StorageManager.sol

```diff
@@ -172,177 +156,181 @@ contract StorageManager {
         require(offer.billingPlans[agreement.billingPeriod] != 0, "StorageManager: price not available anymore");
-        require(msg.value != 0 && msg.value % agreement.billingPrice == 0, "StorageManager: value sent not corresponding to price");
-        //NO_OVERFLOW reasoning: see REF_DURATION
-        require((agreement.startDate + agreement.numberOfPeriodsDeposited) * agreement.billingPeriod <= now, "StorageManager: Agreement expired");
-        //NO_OVERFLOW reasoning: chosenPrice is verified to not be zero in function call: newRequest
-        uint256 numberOfPeriodsDeposited = (msg.value / agreement.size ) / agreement.billingPrice;
-        require(
-            (
-                //NO_OVERFLOW reasoning: billingPeriod can't be zero, startDate always less or equal than now, periodsPast since startDate always less than periodsDeposited (as agreement is not expired, see line 242)
-                agreement.numberOfPeriodsDeposited - (now - agreement.startDate) / agreement.billingPeriod // the amount of periods deposited MINUS periods past since the start date (= periods left)
-                .add(numberOfPeriodsDeposited)
-            ).mul(agreement.billingPeriod) <= offer.maximumDuration &&
-            numberOfPeriodsDeposited.add(agreement.numberOfPeriodsDeposited) <= MAX_UINT64,
-            "StorageManager: period too long");
-        numberOfPeriodsDeposited.mul(agreement.billingPrice); // overFlow check. If this doesn't pass, we might have issues transfering the maximum amount REF_MAX_TRANSFER
-        // NO_OVERFLOW reasoning: verified in line above
-        agreement.numberOfPeriodsDeposited += uint64(numberOfPeriodsDeposited);
-        //NO_OVERFLOW reasoning: verified in function newRequest
-        agreement.startDate.add(agreement.numberOfPeriodsDeposited * agreement.billingPeriod); // overFlow check. If this doesn't pass, the duration of the offer overflows MAX_UINT64 and the contract may deadlock. REF_DURATION
+
+        agreement.availableFunds = agreement.availableFunds.add(msg.value);
         emit AgreementFundsDeposited(agreementReference, msg.value);
     }
 
     /**
-    @notice stops an active agreement.
-    @param dataReference the reference to the to-be-pinned file
+    >> FOR CONSUMER
+    @notice withdraw funds from Agreement.
+    @dev
+        - if amount is zero then all withdrawable funds are transferred (eq. all available funds minus funds for still non-payed out periods and current period)
+        - if Agreement is terminated Consumer can withdraw all remaining funds
+    @param dataReference the data reference of agreement to be funds withdrawn from
     @param provider the address of the provider of the Offer.
     */
-    function terminateAgreement(bytes32[] memory dataReference, address provider) public payable {
+    function withdrawFunds(bytes32[] memory dataReference, address provider, uint256 amount) public payable {
+        Offer storage offer = offerRegistry[provider];
```

Comment:
There needs to be a check if the `msg.sender` is the `agreementAuthor`

Reply:
Nope, see me earlier comment about `agreementReference`. I have refactored the `getAgreementReference()` to be more clear that the `msg.sender` is used for the reference's computation.